cirq-core/cirq/optimizers/two_qubit_to_sqrt_iswap.py

```diff
@@ -0,0 +1,413 @@
+    )
+
+    # Inner single-qubit gates: Fig. 4 of the paper
+    b0 = (
+        protocols.unitary(ops.rz(-gamma))
+        @ protocols.unitary(ops.rx(-alpha))
+        @ protocols.unitary(ops.rz(-gamma))
+    )
+    b1 = protocols.unitary(ops.rx(-beta))
+
+    # Compute KAK on the decomposition to determine outer single-qubit gates
+    # There is no known closed form solution for these gates
+    u_sqrt_iswap = protocols.unitary(ops.SQRT_ISWAP)
+    u = u_sqrt_iswap @ np.kron(b0, b1) @ u_sqrt_iswap  # Unitary of decomposition
+    kak_fix = linalg.kak_decomposition(u, atol=atol / 10, rtol=0, check_preconditions=False)
+    d0, d1 = kak_fix.single_qubit_operations_before
+    e0, e1 = kak_fix.single_qubit_operations_after
+
+    return [  # Pairs of single-qubit unitaries, SQRT_ISWAP between each is implied
+        (d0.T.conj() @ a0, d1.T.conj() @ a1),
+        (b0, b1),
+        (c0 @ e0.T.conj(), c1 @ e1.T.conj()),
+    ], kak.global_phase / kak_fix.global_phase
+
+
+def _decomp_3sqrt_iswap_matrices(
+    kak: 'cirq.KakDecomposition',
+    atol: float = 1e-8,
+) -> Tuple[Sequence[Tuple[np.ndarray, np.ndarray]], complex]:
+    """Returns the single-qubit matrices for the 3-SQRT_ISWAP decomposition.
+
+    Assumes any canonical x, y, z.  Three sqrt-iSWAP gates are only needed if
+    x < y + |z|.  Only two are needed for other gates (most cases).
+
+    References:
+        Towards ultra-high fidelity quantum operations: SQiSW gate as a native
+        two-qubit gate
+        https://arxiv.org/abs/2105.06074
+    """
+    x, y, z = kak.interaction_coefficients
+    f0, f1 = kak.single_qubit_operations_before
+    g0, g1 = kak.single_qubit_operations_after
+
+    # Find x1, y1, z1, x2, y2, z2
+    # such that x1+x2=x, y1+y2=y, z1+z2=z
+    # where x1, y1, z1 are implementable by one sqrt-iSWAP gate
+    # and x2, y2, z2 implementable by two sqrt-iSWAP gates
+    # No error tolerance needed
+    ieq1 = y > np.pi / 8
+    ieq2 = z < 0
+    if ieq1:
+        if ieq2:
+            # Non-canonical Weyl coordinates for the single sqrt-iSWAP
+            x1, y1, z1 = 0, np.pi / 8, -np.pi / 8
+        else:
+            x1, y1, z1 = 0, np.pi / 8, np.pi / 8
+    else:
+        x1, y1, z1 = -np.pi / 8, np.pi / 8, 0
+    # Non-canonical Weyl coordinates for the two sqrt-iSWAP decomposition
+    x2, y2, z2 = x - x1, y - y1, z - z1
+
+    # Find fixup single-qubit gates for the canonical decompositions
+    kak1 = linalg.kak_canonicalize_vector(x1, y1, z1, atol)
+    kak2 = linalg.kak_canonicalize_vector(x2, y2, z2, atol)
+
+    # Compute sub-decompositions
+    ((a0, a1), (b0, b1)), phase1 = _decomp_1sqrt_iswap_matrices(kak1, atol)
+    ((c0, c1), (d0, d1), (e0, e1)), phase2 = _decomp_2sqrt_iswap_matrices(kak2, atol)
+
+    # There are two valid solutions: kak1 before kak2 or kak2 before kak1
+    # Arbitrarily pick kak1 before kak2
+    return [  # Pairs of single-qubit unitaries, SQRT_ISWAP between each is implied
```

Comment:
again, the paper's notation might be good to follow, for easier cross referencing

Reply:
Sure.  Although for the 3-gate case, I didn't quite follow the paper.